library/core/src/task/wake.rs

```diff
@@ -215,6 +283,125 @@ impl fmt::Debug for Context<'_> {
     }
 }
 
+/// A Builder used to construct a `Context` instance
+/// with support for `LocalWaker`.
+///
+/// # Examples
+/// ```
+/// #![feature(local_waker)]
+/// #![feature(noop_waker)]
+/// use std::task::{ContextBuilder, LocalWaker, Waker, Poll};
+/// use std::future::Future;
+///
+/// let local_waker = LocalWaker::noop();
+/// let waker = Waker::noop();
+///
+/// let mut cx = ContextBuilder::from_local_waker(&local_waker)
+///     .waker(&waker)
+///     .build();
+///
+/// let mut future = std::pin::pin!(async { 20 });
+/// let poll = future.as_mut().poll(&mut cx);
+/// assert_eq!(poll, Poll::Ready(20));
+///
+/// ```
+#[unstable(feature = "local_waker", issue = "118959")]
+#[derive(Debug)]
+pub struct ContextBuilder<'a> {
+    waker: Option<&'a Waker>,
+    local_waker: &'a LocalWaker,
+}
+
+impl<'a> ContextBuilder<'a> {
+    /// Create a ContextBuilder from a Waker.
+    #[inline]
+    #[rustc_const_unstable(feature = "const_waker", issue = "102012")]
+    #[unstable(feature = "local_waker", issue = "118959")]
+    pub const fn from_waker(waker: &'a Waker) -> Self {
+        // SAFETY: LocalWaker is just Waker without thread safety
+        let local_waker = unsafe { transmute(waker) };
+        Self { waker: Some(waker), local_waker }
+    }
+
+    /// Create a ContextBuilder from a LocalWaker.
+    #[inline]
+    #[rustc_const_unstable(feature = "const_waker", issue = "102012")]
+    #[unstable(feature = "local_waker", issue = "118959")]
+    pub const fn from_local_waker(local_waker: &'a LocalWaker) -> Self {
+        Self { local_waker, waker: None }
+    }
+
+    /// This field is used to set the value of the waker on `Context`.
+
+    #[inline]
+    #[rustc_const_unstable(feature = "const_waker", issue = "102012")]
+    #[unstable(feature = "local_waker", issue = "118959")]
+    pub const fn waker(self, waker: &'a Waker) -> Self {
+        Self { waker: Some(waker), ..self }
```

Comment:
It feels strange that `from_waker` implicitly sets the local_waker, but `waker(waker)` does not. Can we at least highlight that difference in documentation? While this is unstable surface area it doesn't matter too much, I suppose.

Reply:
We could change it to work like this: 
```rust
 pub fn waker(self, waker: &'a Waker) -> Self {
         let Some(waker) = self.waker else {
            return Self { waker: Some(waker), ..self };
        };

        let override_waker = waker.as_ref().will_wake(self.local_waker);

        if !override_waker {
            return Self { waker: Some(waker), ..self };
        }
        // SAFETY: LocalWaker is just Waker without thread safety
        let local_waker = unsafe { transmute(waker) };
        return Self { waker: Some(waker), local_waker, ..self };
    }
```
This way, the local waker will only be overridden if it hasn't been specialized yet. But then the API would seize to be `const`, unless we are willing to compare them like this: 
```rust
let override_waker = unsafe {
      transmute::<_, [usize; 2]>(waker) != transmute::<_, [usize; 2]>(self.local_waker)
};
```
What do you think it's best?

Reply:
I do think it is unlikely users will set the waker twice, so I wouldn't really mind staying with the original implementation.